types/coin.go

```diff
@@ -664,13 +664,5 @@ func ParseCoins(coinsStr string) (Coins, error) {
 		coins[i] = coin
 	}
 
-	// sort coins for determinism
-	coins.Sort()
-
-	// validate coins before returning
-	if err := coins.Validate(); err != nil {
-		return nil, err
-	}
-
-	return coins, nil
+	return NewCoins(coins...), nil
```

Comment:
This would cause `ParseCoins()` to `panic` if validation fails (`NewCoins()` call `Validate()` yet by design it does not handle panics gracefully). This would quite certainly have repercussions on the client. Are you sure this is intended?

Reply:
Hi @alessio , thanks for asking.
This is intended as I believe that all the Coin validations are already done by `ParseCoin`

![image](https://user-images.githubusercontent.com/30392990/93661534-d1137200-fa58-11ea-8c05-1f482f0a0cea.png)

Similar arguments applied `ParseDecCoins`

Reply:
I've added a test case: https://github.com/cosmos/cosmos-sdk/pull/7348#discussion_r492254520

Reply:
I'd leave the sorting and validation code for `Coins` as a struct but still call `NewCoins` so the coins are still sanitized (removing the zeros)

Reply:
Thinking better,

by adding this validation,

```go
if err := coins.Validate(); err != nil {
		return nil, err
	}
```

back to  `ParseCoins`, the call to `NewCoins` is not necessary anymore because the validation will always throws an err, (`0stake: coin 0stake amount is not positive`), for any zero coin input. See the newly added test cases:

```
{"0stake", false, nil},
{"0stake,1foo,99bar", false, nil},
```
So, in the end, the original implementation of `ParseCoins` was already taking care of zero coins input. 
Shall we revert back to the original implementation and add only the newly test cases to this PR? :smile: 

(Similar logic applied for `ParseDecCoins`)

Reply:
Ah good point!

My 2 satoshis is that `ParseCoins` should mimic functionality of `NewCoins`. `NewCoins` will not panic or return an error when passed a set of coins containing one or more zero coins, I think `ParseCoins` should follow this behaviour OR it should be well documented in the godoc that the expected results differ. I think adding a shared function `sanitizeCoins` called by `NewCoins` and `ParseCoins` which simply removes the zero coins would be sufficient. @alessio thoughts?

Sorry for any confusion, `coins.go` has always be troublesome to make modifications to.

Reply:
> I think ParseCoins should follow this behaviour

I agree. For simplicity's sake, `NewCoins` and `ParseCoins` should have the same `Coins` validation mechanism, with the difference that the latter should return an error whilst the former should just panic in case of errors :+1:

Reply:
@dauTT are you ok with @colin-axner's [proposal](https://github.com/cosmos/cosmos-sdk/pull/7348#discussion_r493315252)? Or you'd rather doing something different?

Reply:
Yes, I like @colin-axner idea.